### PR TITLE
SNOW-670988 Support strict UD(T)F/SP

### PR DESCRIPTION
**Pull request by sfc-gh-sfan**
Description

Supporting strict will allow skip invoking the python function if input contains NULL.

Testing

integ

Please answer these questions before submitting your pull requests. Thanks!

1. What GitHub issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   Fixes SNOW-670988

2. Fill out the following pre-review checklist:

   - [x] I am adding a new automated test(s) to verify correctness of my new code
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency

3. Please describe how your code solves the related issue.

   Supporting strict will allow skip invoking the python function if input contains NULL.

### CODE REVIEW

File: src/snowflake/snowpark/_internal/udf_utils.py [hunk 1]
```diff
@@ -672,11 +673,15 @@ def create_python_udf_or_sp(
         else ""
     )
 
+    strict_as_sql = ""
```
Comment (sfc-gh-stan) [inline src/snowflake/snowpark/_internal/udf_utils.py:676]:
```suggestion
    strict_as_sql = "\nSTRICT" if strict_as_sql else ""
```

File: src/snowflake/snowpark/functions.py [hunk 2]
```diff
@@ -3013,6 +3015,9 @@ def udf(
             The source code is dynamically generated therefore it may not be identical to how the
             `func` is originally defined. The default is ``True``.
             If it is ``False``, source code will not be generated or displayed.
+        strict: Whether the created UDF is strict. A strict UDF will not invoke the UDF if any input is
+            null. Instead, a null value will always be returned for that row. Note that the UDF might
+            still return null for non-null inputs.
```
Comment (sfc-gh-stan) [inline src/snowflake/snowpark/functions.py:3020]: Is the last sentence necessary?

**Review by sfc-gh-stan — APPROVED** [review]
LGTM with one small suggested change.
Otherwise: do we really need indicate that UDF/UDTF/SP can return null for non-null inputs? I feel like that's expected behavior.

File: src/snowflake/snowpark/functions.py (hunk 2 above)
Comment (sfc-gh-stan) [reply src/snowflake/snowpark/functions.py:3020]: Same applied to UDTF and SP docstrings.

Comment (sfc-gh-sfan) [reply src/snowflake/snowpark/functions.py:3020]: This is copied from https://docs.snowflake.com/en/sql-reference/sql/create-function.html#optional-parameters, but I'm open to remove it :)

File: src/snowflake/snowpark/_internal/udf_utils.py [hunk 3]
```diff
@@ -672,11 +673,15 @@ def create_python_udf_or_sp(
         else ""
     )
 
+    strict_as_sql = "\nSTRICT" if strict_as_sql else ""
+    if strict:
+        strict_as_sql = "\nSTRICT"
```
Comment (sfc-gh-stan) [inline src/snowflake/snowpark/_internal/udf_utils.py:678]: Sorry, I probably didn't do the suggestion properly. Could you remove line 677-678 please?

File: src/snowflake/snowpark/functions.py (hunk 2 above)
Comment (sfc-gh-stan) [reply src/snowflake/snowpark/functions.py:3020]: I vote for removal then. @sfc-gh-aling What do you think?

File: src/snowflake/snowpark/_internal/udf_utils.py (hunk 3 above)
Comment (sfc-gh-sfan) [reply src/snowflake/snowpark/_internal/udf_utils.py:678]: Oh yeah let me do that lol

**Review by sfc-gh-aling — APPROVED** [review]
LGTM

File: src/snowflake/snowpark/functions.py (hunk 2 above)
Comment (sfc-gh-aling) [reply src/snowflake/snowpark/functions.py:3020]: though I feel it's not needed, but I think it's a good practice to align with the doc for searchability, and being verbose with one extra sentence doesn't harm too much.

Comment (sfc-gh-sfan) [reply src/snowflake/snowpark/functions.py:3020]: I guess I'll just keep it? :)

**Pull request merged**